internal/proxier/proxy.go

```diff
@@ -64,58 +75,257 @@ type ProxyOpts struct {
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			if err == nil {
+				p.queue.Add(key)
+			}
+		},
+	})
+
+	endpointsInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(oldObj, obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err == nil {
+				p.queue.Add(key)
+			}
+		},
+	})
+	return p, endpointsInformer.AddIndexers(cache.Indexers{
+		"pods": IndexEndpointsByActivePod,
+	})
 }
 
 // Start starts the proxier
 // TODO: replace raw cluster domain with options struct, maybe also
 // move into NewProxier
 func (p *Proxier) Start(ctx context.Context) error {
+	defer p.queue.ShutDown()
+
 	log := p.log.WithField("component", "proxier")
 	portForwarder, pfdoneChan, worker, err := NewPortForwarder(ctx, p.k, p.rest, p.log, p.opts)
+	p.pfrequest = portForwarder
+
+	log.Infof("Starting %d proxier worker(s)", p.threadiness)
+	for i := 0; i < p.threadiness; i++ {
+		go wait.Until(p.runWorker, time.Second, ctx.Done())
+	}
+
 	if err != nil {
 		return err
 	}
 	p.worker = worker
 
-	serviceChan, handlerDoneChan := CreateHandlers(ctx, portForwarder, p.k, p.opts.ClusterDomain)
+	<-ctx.Done()
+	log.Info("waiting for port-forward worker to finish")
+	<-pfdoneChan
+	return nil
+}
+
+func (p *Proxier) runWorker() {
+	for p.processNextWorkItem() {
+
+	}
+}
+
+func (p *Proxier) processNextWorkItem() bool {
+	key, quit := p.queue.Get()
+	if quit {
+		return false
+	}
+	defer p.queue.Done(key)
+
+	// Invoke the method containing the business logic
+	err := p.reconcile(key.(string))
+
+	if err == nil {
+		// Forget about the #AddRateLimited history of the key on every successful synchronization.
+		// This ensures that future processing of updates for this key is not delayed because of
+		// an outdated error history.
+		p.queue.Forget(key)
+		return true
+	}
+
+	if p.queue.NumRequeues(key) < 5 {
+		// Re-enqueue the key rate limited. Based on the rate limiter on the
+		// queue and the re-enqueue history, the key will be processed later again.
+		p.queue.AddRateLimited(key)
+		return true
+	}
+
+	// Retries exceeded. Forgetting for this reconciliation loop
+	p.queue.Forget(key)
+	return true
+}
 
-	err = kevents.WaitForSync(ctx, kevents.GlobalCache.TrackObject("endpoints", &corev1.Endpoints{}))
+func (p *Proxier) reconcile(key string) error {
+	o, exists, err := p.svcInformer.GetStore().GetByKey(key)
 	if err != nil {
-		return errors.Wrap(err, "failed to sync endpoint cache")
+		return err
 	}
 
-	// Handle services being created, send them to the proxier
-	err = kevents.GlobalCache.Subscribe(ctx, "services", &corev1.Service{}, func(e kevents.Event) {
-		if e.Event == kevents.EventTypeUpdated {
-			return
+	if !exists {
+		//we don')t have the service object anymore, we need to get the namespace/name from the key
+		namespace, name, err := cache.SplitMetaNamespaceKey(key)
+		if err != nil {
+			return err
+		}
+		p.pfrequest <- PortForwardRequest{
```

Comment:
Should we check if we're tracking this service before trying to delete it? I think this would solve the issue I posted earlier.

Reply:
Alternatively we can just.... not log the error. Since, I don't think it's an issue.

Reply:
We could go with ways. It comes done to the question if we want these request to be indempotent or if its only allowed to request the deletion of an actually existing portforward. The create request is idempotent and exists early if there is already a port forward recorded so I'm leaning towards doing it the same here and make the "api" simpler in that the caller does not need to care.

Reply:
I opted for making  `DeletePortForward` skip silently over non exiting port forwards...